ruamel-examples/parse_it.py

```diff
@@ -0,0 +1,237 @@
+                return rval
+            else:
+                raise ValueError(f"Environment variable {env_variable} not found")
+        rval = super().construct_scalar(node)
+
+        return rval
+
+
+class ProvenanceConstructor(EnvironmentConstructor):
+    """
+    Subclasses the ``EnvironmentConstructor`` to, instead of returning only a ``data``
+    returning a ``tuple`` where the element 0 is the ``data`` itself and the following
+    elements are the provenance values (1 -> line number, 2 -> column). The resulting
+    dictionary contains keys that are ``tuples`` and ``values`` that are tuples. This
+    can then be separated into two equivalent dictionaries, one with the "real" values
+    and another one with the values of the provenance.
+    """
+
+    def construct_object(self, node, *args, **kwargs):
+        """
+        Parameters
+        ----------
+        node : node object
+            The node containing all the information about the yaml element
+
+        Returns
+        -------
+        data : any
+            The "real" value as interpreted from the parent yaml constructor
+        provenance : tuple
+            provenance[0]: line number
+            provenance[1]: column number
+        """
+
+        data = super().construct_object(node, *args, **kwargs)
+
+        provenance = (
+            node.start_mark.line,
+            node.start_mark.column,
+        )
+
+        return (data, provenance)
+
+
+class EnvironmentRepresenter(ruamel.yaml.RoundTripRepresenter):
+    """When dumping, this class is used to remove the !ENV tag from a particular value."""
+
+    def represent_scalar(self, tag, value, style=None, anchor=None):
+        type_tags = {
+            "int": f"tag:{YAML_SPEC_TAG}:int",
+            "float": f"tag:{YAML_SPEC_TAG}:float",
+            "bool": f"tag:{YAML_SPEC_TAG}:bool",
+            "str": f"tag:{YAML_SPEC_TAG}:str",
+        }
+        if tag == "!ENV":
+            type_tag = type(value).__name__
+            actual_type = type_tags[type_tag]
+            return super().represent_scalar(actual_type, value, style, anchor=anchor)
+        return super().represent_scalar(tag, value, style, anchor=anchor)
+
+
+class EsmToolsLoader(ruamel.yaml.YAML):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.filename = None
+        self.add_comments = True
+        self.Constructor = ProvenanceConstructor
+        self.Constructor.add_constructor(None, ProvenanceConstructor.construct_scalar)
+        self.Constructor.add_constructor(
+            "tag:yaml.org,2002:bool", ProvenanceConstructor.construct_yaml_bool
+        )
+        self.Representer = EnvironmentRepresenter
+
+    def get_filename(self):
+        return self.filename
+
+    def set_filename(self, filename):
+        self.filename = filename
+
+    def load(self, stream):
+        self.set_filename(stream.name)
+        mapping_with_tuple_prov = super().load(stream)
```

Comment:
I would recommend extracting already at this point the first level dictionary, instead of it being a tuple:

```python3
mapping_with_tuple_prov = super().load(stream)[0]
```

Reply:
Just for clarification:
This `mapping_with_tuple_prov = super().load(stream)` returns a tuple of length 2.
`mapping_with_tuple_prov[0]` is of type `<class 'ruamel.yaml.comments.CommentedMap'>` and 
`type(mapping_with_tuple_prov[1]` of type `<class 'tuple'>`

```
        print(f"{type(mapping_with_tuple_prov)} = {mapping_with_tuple_prov}")
        print()
        print(f"{type(mapping_with_tuple_prov[0])} = {mapping_with_tuple_prov[0]}")
        print()
        print(f"{type(mapping_with_tuple_prov[1])} = {mapping_with_tuple_prov[1]}")
        print()
        print(f"length: {len(mapping_with_tuple_prov)}")
```

returns

```
<class 'tuple'> = ({('person', (2, 0)): ({('name', (3, 2)): ('Paul Gierz', (3, 8)), ('username', (4, 2)): ('nwieters', (4, 12)), ('a_string', (5, 2)): (' hello world I am here to make your life impossible ', (5, 12)), ('my_var', (10, 2)): ('TEST', (10, 10)), ('my_other_var', (11, 2)): ([('a', (12, 8)), ('b', (14, 8)), ('c', (15, 8))], (12, 6)), ('my_other_list', (16, 2)): ([('a', (16, 18)), ('b', (16, 21)), ('c', (16, 24))], (16, 17)), ('my_bolean', (18, 2)): (True, (18, 13)), ('my_int', (19, 2)): (12.1, (19, 10)), ('my_int2', (20, 2)): (44, (20, 11))}, (3, 2))}, (2, 0))

<class 'ruamel.yaml.comments.CommentedMap'> = {('person', (2, 0)): ({('name', (3, 2)): ('Paul Gierz', (3, 8)), ('username', (4, 2)): ('nwieters', (4, 12)), ('a_string', (5, 2)): (' hello world I am here to make your life impossible ', (5, 12)), ('my_var', (10, 2)): ('TEST', (10, 10)), ('my_other_var', (11, 2)): ([('a', (12, 8)), ('b', (14, 8)), ('c', (15, 8))], (12, 6)), ('my_other_list', (16, 2)): ([('a', (16, 18)), ('b', (16, 21)), ('c', (16, 24))], (16, 17)), ('my_bolean', (18, 2)): (True, (18, 13)), ('my_int', (19, 2)): (12.1, (19, 10)), ('my_int2', (20, 2)): (44, (20, 11))}, (3, 2))}

<class 'tuple'> = (2, 0)

length: 2
```

`mapping_with_tuple_prov = super().load(stream)[0]` will work since the tuple of `mapping_with_tuple_prov[1]` is not needed/redundant (provenance of the very first key 'person').

Reply:
I see! Thanks for checking that out :)